### PR TITLE
enable AAD token authentication

### PR DESCRIPTION
**Pull request by rogerpengyu**
(https://github.com/prisma/tiberius/issues/175)

### CODE REVIEW

File: src/client/connection.rs [hunk 1]
```diff
@@ -81,7 +81,14 @@ impl<S: AsyncRead + AsyncWrite + Unpin + Send> Connection<S> {
             buf: BytesMut::new(),
         };
 
-        let prelogin = connection.prelogin(config.encryption).await?;
+        let fed_auth_required = if let AuthMethod::AADToken(_) = config.auth {
+            true
+        } else {
+            false
+        };
+        let prelogin = connection
```
Comment (pimeys) [inline src/client/connection.rs:89]: whitespace before this line

File: src/tds/codec/login.rs [hunk 2]
```diff
@@ -201,6 +249,20 @@ impl<'a> LoginMessage<'a> {
...
     }
+
+    pub fn aad_token(
+        &mut self,
+        token: impl Into<Cow<'a, str>>,
+        fed_auth_echo: bool,
+        nonce: Option<[u8; 32]>,
+    ) {
+        self.option_flags_3.insert(OptionFlag3::ExtensionUsed);
+        self.fed_auth_ext = Some(FedAuthExt {
```
Comment (pimeys) [inline src/tds/codec/login.rs:260]: whitespace before this line

File: src/tds/codec/token/token_feature_ext_ack.rs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#[derive(Debug)]
+pub struct TokenFeatureExtAck {
+    pub features: Vec<FeatureAck>,
+}
+
+#[derive(Debug)]
+pub enum FedAuthAck {
+    SecurityToken { nonce: Option<[u8; 32]> },
+}
+#[derive(Debug)]
```
Comment (pimeys) [inline src/tds/codec/token/token_feature_ext_ack.rs:13]: whitespace before this line

**Review by pimeys — CHANGES_REQUESTED** [review]
One thing missing: JDBC and ASP.net connection string parameters. Should be whatever Azure is using.

**Comment by rogerpengyu** [comment]
> One thing missing: JDBC and ASP.net connection string parameters. Should be whatever Azure is using.

the connection string usually specifies AAD user name and secrets. Then it talks to AAD to retrieve a token. If we code that functionality in tiberious, we will need to include a http lib for token retrieval. I am trying to avoid that by just specifying the retrieved token. User could call AAD with their own code to retrieve the token. And call Config.aad_token() directly.

Let me know if you want to include the AAD token retrieval in tiberius

**Comment by pimeys** [comment]
How do you get the token in Azure? I'm thinking about a public interface for Prisma users, who mainly configure SQL Server with the JDBC connection string. Can you get the token easily from Azure and how would it work then together with Tiberius?

**Comment by rogerpengyu** [comment]
ODBC connection string could support (1) AAD user name + password; or (2) AAD service principal + secret. Both will take user id + secret from connection string, then making a REST API call to AAD to retrieve token, and pass the token to SQL server.

Calling AAD to retrieve token is straightforward, but it requires a HTTP client. Below is the code I used to retrieve token:

`        let token_body: Vec<u8> = form_urlencoded::Serializer::new(String::new())
            .append_pair("grant_type", "client_credentials")
            .append_pair("client_id", self.client_id.as_str())
            .append_pair("client_secret", self.client_secret.as_str())
            .append_pair("resource", self.resource_url.as_str())
            .finish()
            .into_bytes();
        let token_request = new_request()
            .method(Method::POST)
            .uri(format!("{}/{}/oauth2/token", self.authority_url, self.tenant_id))
            .header(HEADER_CONTENT_TYPE, CONTENT_TYPE_FORM_URLENCODED)
            .body(token_body)
            .expect("build token retrieval request should succeed");

        let res = self.http_client.clone().request(token_request.into())?.success()?;

        let token_body = String::from_utf8(res.into_body().to_vec()).expect("token response should always be utf8 encoded");
        let kvs: HashMap<String, String> = serde_json::from_str(token_body.as_str()).expect("token response should be valid json");

        let token = BearerToken(kvs["access_token"].to_string());
        let expired_on: Option<i64> = kvs.get("expires_on").map(|s| s.as_str().parse().unwrap());
`

There is not standard way to specify token in connection string directly. but one can specify the token in the properties for other sql client library. See https://docs.microsoft.com/en-us/sql/connect/jdbc/setting-the-connection-properties?view=sql-server-ver15

"accessTokenStringnull | (Version 6.0+) Use this property to connect to a database using an access token. accessToken can't be set using the connection URL."

And this is how the lib works in java:

`        IAuthenticationResult authenticationResult = future.get();
        String accessToken = authenticationResult.accessToken();

        System.out.println("Access Token: " + accessToken);

        // Connect with the access token.
        SQLServerDataSource ds = new SQLServerDataSource();

        ds.setServerName("aad-managed-demo.database.windows.net"); // Replace with your server name.
        ds.setDatabaseName("demo"); // Replace with your database name.
        ds.setAccessToken(accessToken);`

So there are two options:
- enable userid/password in connection string, and include a HTTP client in tiberius;
- only supports token in tiberius, which cannot be specified by another interface than connection string

**Comment by pimeys** [comment]
I think from the user perspective a HTTP client would make sense. The question is would it lock us to Tokio for good? Or could it be built so that we can still support other runtimes too?

I'm also thinking of a separate crate, such as `tiberius-aad` for the token retrieval. This makes sense when looking into the highly modular architecture of Tiberius, but we'd still need the connection string parsing to work, which might be tricky if the AAD stuff is in another crate.

**Comment by rogerpengyu** [comment]
if we are ok to include a HTTP client, we could use reqwest, which is a synchronized lib (a good fit for your scenario as we only made very few http requests). it won't lock tiberius to async runtime as well.

let me know which way you want me to move. Either option (keep token only, or add reqwest to support password) will be ok to me.

**Comment by rogerpengyu** [comment]
so what's the suggestion here?

**Comment by rogerpengyu** [comment]
The azure-identity crate is just making the rest api call (see the code snippets I pasted above). If we all agree to include the token retrieval in Tiberius, that's one option.

@pimeys could you make the call here? Do you suggest to keep token retrieval in Tiberius (in which case we will depend on one http crate or azure identity crate)?

Or we can say Tiberius only support token as interface. Users are suggested to use azure-identify to retrieve token themselves.

**Comment by pimeys** [comment]
Latter. But we need an example how to use it and docs.

**Comment by rogerpengyu** [comment]
Sure. To be clear, that means one cannot use connect string to use AAD. They have to call the Config api to setup. Are you ok with that?

**Comment by rogerpengyu** [comment]
on the doc string I add an example for token retrieval. @pimeys please take a look

**Comment by rogerpengyu** [comment]
@pimeys : any other suggestions for the PR?

**Comment by rogerpengyu** [comment]
ping @pimeys

**Comment by pimeys** [comment]
On vacation. Back next week.

File: src/tds/codec/pre_login.rs [hunk 4]
```diff
@@ -76,38 +78,61 @@ impl PreloginMessage {
...
-                dst.put_u16(data_offset);
-                dst.put_u16(length);
+        // version
+        fields.push((PRELOGIN_VERSION, 0x04 + 0x02)); // version + subbuild
+        data_cursor.write_u32::<BigEndian>(self.version as u32)?;
+        data_cursor.write_u16::<BigEndian>(self.sub_build as u16)?;
 
-                data_offset += length;
+        // encryption
+        fields.push((PRELOGIN_ENCTYPTION, 0x01)); // encryption
```
Comment (pimeys) [inline src/tds/codec/pre_login.rs:104]:
```suggestion
        fields.push((PRELOGIN_ENCRYPTION, 0x01)); // encryption
```

File: src/tds/codec/token/token_feature_ext_ack.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        loop {
+            let feature_id = src.read_u8().await?;
+
+            if feature_id == FEA_EXT_TERMINATOR {
+                break;
+            } else if feature_id == FEA_EXT_FEDAUTH {
+                let data_len = src.read_u32_le().await?;
+                let nonce = if data_len == 32 {
+                    let mut n = [0u8; 32];
+                    src.read_exact(&mut n).await?;
```
Comment (pimeys) [inline src/tds/codec/token/token_feature_ext_ack.rs:34]:
```suggestion
                    src.read_exact(&mut n).await?;

```

File: src/tds/codec/token/token_feature_ext_ack.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                    Some(n)
+                } else if data_len == 0 {
+                    None
+                } else {
+                    panic!("invalid Feature_Ext_Ack token");
+                };
```
Comment (pimeys) [inline src/tds/codec/token/token_feature_ext_ack.rs:40]:
```suggestion
                };

```

File: src/tds/codec/login.rs [hunk 7]
```diff
@@ -128,8 +129,54 @@ pub enum LoginTypeFlag {
     ReadOnlyIntent = 1 << 5,
 }
 
+// pub const SESSIONRECOVERY: u8 = 0x01u8;
```
[the remark this answers is not in the document]
Comment (rogerpengyu) [reply src/tds/codec/login.rs:132]: taken

File: src/tds/codec/login.rs [hunk 8]
```diff
@@ -128,8 +129,54 @@ pub enum LoginTypeFlag {
...
+pub const FEA_EXT_FEDAUTH: u8 = 0x02u8;
```
[the remark this answers is not in the document]
Comment (rogerpengyu) [reply src/tds/codec/login.rs:133]: taken

File: src/tds/codec/pre_login.rs (hunk 4 above)
Comment (rogerpengyu) [reply src/tds/codec/pre_login.rs:104]: taken

File: src/tds/codec/login.rs [hunk 9]
```diff
@@ -317,3 +423,215 @@ impl<'a> Encode<BytesMut> for LoginMessage<'a> {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
```
[the remark this answers is not in the document]
Comment (rogerpengyu) [reply src/tds/codec/login.rs:421]: I will suggest we keep this. It's difficult to setup integrated aad auth test. These are the only protection layer for code quality.

[the remark this answers is not in the document]
Comment (rogerpengyu) [reply src/tds/codec/login.rs:396]: you don't have to read all the lines (as these are tests code), pretty much it runs a loopback test for serde and to check serialization could be correctly deserialized.

**Comment by rogerpengyu** [comment]
all comments addressed. please take another look

File: examples/aad-auth.rs [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    .await?;
+
+    let mut config = Config::new();
+    let server = env::var("SERVER").expect("Missing HOST environment variable.");
+    config.host(server);
+    config.port(1433);
+    config.authentication(AuthMethod::AADToken(token.access_token().secret().clone()));
+    config.trust_cert();
+    let tcp = TcpStream::connect(config.get_addr()).await?;
+    tcp.set_nodelay(true)?;
```
Comment (pimeys) [inline examples/aad-auth.rs:46]:
```suggestion
    tcp.set_nodelay(true)?;

```

**Review by pimeys — CHANGES_REQUESTED** [review]
Few nitpicks, but the tests must be green before merging.

**Comment by pimeys** [comment]
Thank you, I'll cut a release.

**Pull request merged**